File: csl-rfid-android-sdk/src/main/java/com/csl/rfidsdk/models/RfidConfiguration.java
package com.csl.rfidsdk.models;

import com.csl.rfidsdk.config.RfidInventoryMode;
import com.csl.rfidsdk.config.RfidRegion;
import com.csl.rfidsdk.config.RfidTarget;

/**
 * Configuration settings for RFID operations
 */
public class RfidConfiguration {
    private final int powerLevel;
    private final int session;
    private final int qValue;
    private final RfidTarget target;
    private final RfidInventoryMode inventoryMode;
    private final RfidRegion region;
    private final boolean populateRssi;
    private final boolean populatePhase;
    private final boolean populateChannel;
    private final boolean enableBeep;
    private final boolean enableVibrate;

    private RfidConfiguration(Builder builder) {
        this.powerLevel = builder.powerLevel;
        this.session = builder.session;
        this.qValue = builder.qValue;
        this.target = builder.target;
        this.inventoryMode = builder.inventoryMode;
        this.region = builder.region;
        this.populateRssi = builder.populateRssi;
        this.populatePhase = builder.populatePhase;
        this.populateChannel = builder.populateChannel;
        this.enableBeep = builder.enableBeep;
        this.enableVibrate = builder.enableVibrate;
    }

    public int getPowerLevel() {
        return powerLevel;
    }

    public int getSession() {
        return session;
    }

    public int getQValue() {
        return qValue;
    }

    public RfidTarget getTarget() {
        return target;
    }

    public RfidInventoryMode getInventoryMode() {
        return inventoryMode;
    }

    public RfidRegion getRegion() {
        return region;
    }

    public boolean isPopulateRssi() {
        return populateRssi;
    }

    public boolean isPopulatePhase() {
        return populatePhase;
    }

    public boolean isPopulateChannel() {
        return populateChannel;
    }

    public boolean isEnableBeep() {
        return enableBeep;
    }

    public boolean isEnableVibrate() {
        return enableVibrate;
    }

    public static class Builder {
        private int powerLevel = 300; // Default 30.0 dBm
        private int session = 1;
        private int qValue = 4;
        private RfidTarget target = RfidTarget.A;
        private RfidInventoryMode inventoryMode = RfidInventoryMode.COMPACT;
        private RfidRegion region = RfidRegion.FCC;
        private boolean populateRssi = true;
        private boolean populatePhase = false;
        private boolean populateChannel = false;
        private boolean enableBeep = true;
        private boolean enableVibrate = true;

        public Builder powerLevel(int powerLevel) {
            this.powerLevel = powerLevel;
            return this;
        }

        public Builder session(int session) {
            this.session = session;
            return this;
        }

        public Builder qValue(int qValue) {
            this.qValue = qValue;
            return this;
        }

        public Builder target(RfidTarget target) {
            this.target = target;
            return this;
        }

        public Builder inventoryMode(RfidInventoryMode inventoryMode) {
            this.inventoryMode = inventoryMode;
            return this;
        }

        public Builder region(RfidRegion region) {
            this.region = region;
            return this;
        }

        public Builder populateRssi(boolean populateRssi) {
            this.populateRssi = populateRssi;
            return this;
        }

        public Builder populatePhase(boolean populatePhase) {
            this.populatePhase = populatePhase;
            return this;
        }

        public Builder populateChannel(boolean populateChannel) {
            this.populateChannel = populateChannel;
            return this;
        }

        public Builder enableBeep(boolean enableBeep) {
            this.enableBeep = enableBeep;
            return this;
        }

        public Builder enableVibrate(boolean enableVibrate) {
            this.enableVibrate = enableVibrate;
            return this;
        }

        public RfidConfiguration build() {
            return new RfidConfiguration(this);
        }
    }
}

File: cs710flutterapp/lib/models/rfid_configuration.dart
/// RFID reader configuration settings
class RfidConfiguration {
  final int powerLevel; // 0-320 (0.0-32.0 dBm)
  final int session; // 0-3
  final int qValue; // 0-15
  final String target; // "A", "B", "AB_FLIP"
  final String inventoryMode; // "STANDARD", "COMPACT"
  final String region; // "FCC", "ETSI", "CN", etc.
  final bool enableBeep;
  final bool enableVibrate;

  const RfidConfiguration({
    required this.powerLevel,
    required this.session,
    required this.qValue,
    required this.target,
    required this.inventoryMode,
    required this.region,
    required this.enableBeep,
    required this.enableVibrate,
  });

  /// Create a builder for fluent configuration
  static RfidConfigurationBuilder builder() {
    return RfidConfigurationBuilder();
  }

  /// Create default configuration (matches cs710aquickstart defaults)
  factory RfidConfiguration.defaultConfig() {
    return const RfidConfiguration(
      powerLevel: 300, // 30.0 dBm
      session: 1,
      qValue: 7,
      target: 'A',
      inventoryMode: 'COMPACT',
      region: 'FCC',
      enableBeep: true,
      enableVibrate: true,
    );
  }

  /// Create RfidConfiguration from platform channel map
  factory RfidConfiguration.fromMap(Map<String, dynamic> map) {
    return RfidConfiguration(
      powerLevel: map['powerLevel'] as int,
      session: map['session'] as int,
      qValue: map['qValue'] as int,
      target: map['target'] as String,
      inventoryMode: map['inventoryMode'] as String,
      region: map['region'] as String,
      enableBeep: map['enableBeep'] as bool,
      enableVibrate: map['enableVibrate'] as bool,
    );
  }

  /// Convert to map for platform channel
  Map<String, dynamic> toMap() {
    return {
      'powerLevel': powerLevel,
      'session': session,
      'qValue': qValue,
      'target': target,
      'inventoryMode': inventoryMode,
      'region': region,
      'enableBeep': enableBeep,
      'enableVibrate': enableVibrate,
    };
  }

  /// Get power level in dBm (e.g., 300 -> 30.0)
  double get powerDbm => powerLevel / 10.0;

  @override
  String toString() =>
      'RfidConfiguration(power: ${powerDbm}dBm, session: $session, target: $target)';
}

/// Builder for RfidConfiguration with fluent API
/// Usage: RfidConfiguration.builder()
///           .powerLevel(300)
///           .session(1)
///           .build()
class RfidConfigurationBuilder {
  int _powerLevel = 300; // Default 30.0 dBm
  int _session = 1;
  int _qValue = 7;
  String _target = 'A';
  String _inventoryMode = 'COMPACT';
  String _region = 'FCC';
  bool _enableBeep = true;
  bool _enableVibrate = true;

  RfidConfigurationBuilder();

  /// Set power level (0-320, representing 0.0-32.0 dBm)
  RfidConfigurationBuilder powerLevel(int value) {
    _powerLevel = value;
    return this;
  }

  /// Set session (0-3)
  RfidConfigurationBuilder session(int value) {
    _session = value;
    return this;
  }

  /// Set Q value (0-15)
  RfidConfigurationBuilder qValue(int value) {
    _qValue = value;
    return this;
  }

  /// Set target ("A", "B", "AB_FLIP")
  RfidConfigurationBuilder target(String value) {
    _target = value;
    return this;
  }

  /// Set inventory mode ("STANDARD", "COMPACT")
  RfidConfigurationBuilder inventoryMode(String value) {
    _inventoryMode = value;
    return this;
  }

  /// Set region ("FCC", "ETSI", "CN", etc.)
  RfidConfigurationBuilder region(String value) {
    _region = value;
    return this;
  }

  /// Enable or disable beep
  RfidConfigurationBuilder enableBeep(bool value) {
    _enableBeep = value;
    return this;
  }

  /// Enable or disable vibrate
  RfidConfigurationBuilder enableVibrate(bool value) {
    _enableVibrate = value;
    return this;
  }

  /// Build the configuration
  RfidConfiguration build() {
    return RfidConfiguration(
      powerLevel: _powerLevel,
      session: _session,
      qValue: _qValue,
      target: _target,
      inventoryMode: _inventoryMode,
      region: _region,
      enableBeep: _enableBeep,
      enableVibrate: _enableVibrate,
    );
  }
}
